lib/open_project/rate_limiting/base.rb

```diff
@@ -0,0 +1,102 @@
+module OpenProject
+  module RateLimiting
+    class Base
+      class << self
+        def rule_name
+          name.demodulize.underscore
+        end
+
+        def enabled?
+          if settings.key?(:enabled)
+            ActiveRecord::Type::Boolean.new.cast(settings[:enabled])
+          else
+            enabled_by_default?
+          end
+        end
+
+        def enabled_by_default?
+          true
+        end
+
+        def settings
+          value = OpenProject::Configuration.rate_limiting.with_indifferent_access.fetch(rule_name, {})
+          value = { enabled: value } if [true, false].include?(value)
+          value
+        end
+      end
+
+      def apply!
+        Rack::Attack.throttle(rule_name, limit:, period:) do |request|
+          limit_condition(request)
+        end
+
+        self
+      end
+
+      delegate :settings, :rule_name, to: :class
+
+      def response(request)
+        match_data = request.env['rack.attack.match_data']
+        now = match_data[:epoch_time]
+        retry_after = match_data[:period] - (now % match_data[:period])
+
+        headers = {
+          'RateLimit-Limit' => match_data[:limit].to_s,
+          'RateLimit-Remaining' => '0',
+          'RateLimit-Reset' => (now + (match_data[:period] - (now % match_data[:period]))).to_s
+        }
+
+        body = response_body(request:, now:, retry_after:)
+        [429, headers, [body]]
+      end
+
+      def response_body(retry_after:, **)
+        "Your request has been throttled. Try again at #{retry_after.seconds.from_now}.\n"
+      end
+
+      protected
+
+      def limit
```

Comment:
I would suggest documenting here that limit (and period too) can be a block that takes the request. This way we can adjust the limit dynamically. In saas this will be required to implement per-tenant limits.

Here's an example from the docs:

```
limit = proc { |req| req.env["REMOTE_USER"] == "admin" ? 100 : 1 }
```

Reply:
I've documented it to prevent calling a proc for limit if it's not necessary